src/nbl/video/IGPUCommandBuffer.cpp

```diff
@@ -0,0 +1,951 @@
+}
+
+bool IGPUCommandBuffer::bindVertexBuffers(uint32_t firstBinding, uint32_t bindingCount, const buffer_t* const* const pBuffers, const size_t* pOffsets)
+{
+    for (uint32_t i = 0u; i < bindingCount; ++i)
+    {
+        if (pBuffers[i] && !this->isCompatibleDevicewise(pBuffers[i]))
+            return false;
+    }
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CBindVertexBuffersCmd>(m_segmentListHeadItr, m_segmentListTail, firstBinding, bindingCount, pBuffers))
+        return false;
+
+    bindVertexBuffers_impl(firstBinding, bindingCount, pBuffers, pOffsets);
+
+    return true;
+}
+
+bool IGPUCommandBuffer::dispatchIndirect(const buffer_t* buffer, size_t offset)
+{
+    if (!buffer || buffer->getAPIType() != getAPIType())
+        return false;
+
+    if (!this->isCompatibleDevicewise(buffer))
+        return false;
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CDispatchIndirectCmd>(m_segmentListHeadItr, m_segmentListTail, core::smart_refctd_ptr<const IGPUBuffer>(buffer)))
+        return false;
+
+    return dispatchIndirect_impl(buffer, offset);
+}
+
+bool IGPUCommandBuffer::setEvent(event_t* _event, const SDependencyInfo& depInfo)
+{
+    if (!_event || _event->getAPIType() != getAPIType())
+        return false;
+
+    if (!this->isCompatibleDevicewise(_event))
+        return false;
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CSetEventCmd>(m_segmentListHeadItr, m_segmentListTail, core::smart_refctd_ptr<const IGPUEvent>(_event)))
+        return false;
+
+    return setEvent_impl(_event, depInfo);
+}
+
+bool IGPUCommandBuffer::resetEvent(event_t* _event, asset::E_PIPELINE_STAGE_FLAGS stageMask)
+{
+    if (!_event || _event->getAPIType() != getAPIType())
+        return false;
+
+    if (!this->isCompatibleDevicewise(_event))
+        return false;
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CResetEventCmd>(m_segmentListHeadItr, m_segmentListTail, core::smart_refctd_ptr<const IGPUEvent>(_event)))
+        return false;
+
+    return resetEvent_impl(_event, stageMask);
+}
+
+bool IGPUCommandBuffer::waitEvents(uint32_t eventCount, event_t* const* const pEvents, const SDependencyInfo* depInfo)
+{
+    if (eventCount == 0u)
+        return false;
+
+    for (uint32_t i = 0u; i < eventCount; ++i)
+    {
+        if (!pEvents[i] || !this->isCompatibleDevicewise(pEvents[i]))
+            return false;
+    }
+
+    constexpr uint32_t MaxBarrierCount = 100u;
+    assert(depInfo->memBarrierCount <= MaxBarrierCount);
+    assert(depInfo->bufBarrierCount <= MaxBarrierCount);
+    assert(depInfo->imgBarrierCount <= MaxBarrierCount);
+
+    const IGPUBuffer* buffers_raw[MaxBarrierCount];
+    for (auto i = 0; i < depInfo->bufBarrierCount; ++i)
+        buffers_raw[i] = depInfo->bufBarriers[i].buffer.get();
+
+    const IGPUImage* images_raw[MaxBarrierCount];
+    for (auto i = 0; i < depInfo->imgBarrierCount; ++i)
+        images_raw[i] = depInfo->imgBarriers[i].image.get();
```

Comment:
same as https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files#r1047477901 but different command

Reply:
Lets tackle in #441